src/Castle.Services.Logging.NLogIntegration/NLogLogger.cs

```diff
@@ -113,13 +122,106 @@ public virtual Core.Logging.ILogger CreateChildLogger(String name)
+		/// <summary>
+		///   Logs a trace message.
+		/// </summary>
+		/// <param name="message"> The message to log </param>
+		public void Trace(string message)
+		{
+			Log(LogLevel.Trace, message);
+		}
+
+		/// <summary>
+		///   Logs a trace message.
+		/// </summary>
+		/// <param name="messageFactory"> Factory constructing lazily the message to log if the level is enabled </param>
+		public void Trace(Func<string> messageFactory)
+		{
+			if (IsTraceEnabled)
+			{
+				Log(LogLevel.Debug, messageFactory());
+			}
+		}
+
+		/// <summary>
+		///   Logs a trace message.
+		/// </summary>
+		/// <param name="exception"> The exception to log </param>
+		/// <param name="message"> The message to log </param>
+		public void Trace(string message, Exception exception)
+		{
+			if (IsTraceEnabled)
+			{
+				Log(LogLevel.Trace, message, exception);
+			}
+		}
+
+		/// <summary>
+		///   Logs a trace message.
+		/// </summary>
+		/// <param name="format"> Format string for the message to log </param>
+		/// <param name="args"> Format arguments for the message to log </param>
+		public void TraceFormat(string format, params object[] args)
+		{
+			if (IsTraceEnabled)
+			{
+				Log(LogLevel.Trace, format, args);
+			}
+		}
+
+		/// <summary>
+		///   Logs a trace message.
+		/// </summary>
+		/// <param name="exception"> The exception to log </param>
+		/// <param name="format"> Format string for the message to log </param>
+		/// <param name="args"> Format arguments for the message to log </param>
+		public void TraceFormat(Exception exception, string format, params object[] args)
+		{
+			if (IsTraceEnabled)
+			{
+				Log(LogLevel.Trace, exception, format, args);
+			}
+		}
+
+		/// <summary>
+		///   Logs a trace message.
+		/// </summary>
+		/// <param name="formatProvider"> The format provider to use </param>
+		/// <param name="format"> Format string for the message to log </param>
+		/// <param name="args"> Format arguments for the message to log </param>
+		public void TraceFormat(IFormatProvider formatProvider, string format, params object[] args)
+		{
+			if (IsTraceEnabled)
+			{
+				Log(LogLevel.Trace, formatProvider, format, args);
+			}
+		}
+
+		/// <summary>
+		///   Logs a trace message.
+		/// </summary>
+		/// <param name="exception"> The exception to log </param>
+		/// <param name="formatProvider"> The format provider to use </param>
+		/// <param name="format"> Format string for the message to log </param>
+		/// <param name="args"> Format arguments for the message to log </param>
+		public void TraceFormat(Exception exception, IFormatProvider formatProvider, string format, params object[] args)
+		{
+			if (IsTraceEnabled)
+			{
+				Log(LogLevel.Trace, exception, formatProvider, format, args);
+			}
+		}
+
 		/// <summary>
 		///   Logs a debug message.
 		/// </summary>
 		/// <param name="message"> The message to log </param>
 		public void Debug(string message)
 		{
-			Log(LogLevel.Debug, message);
+			if (IsDebugEnabled)
```

Comment:
Why did you add all these `if (IsDebugEnabled)` type checks to existing log methods, the `IsXXXEnabled` check is only needed on log methods taking a `Func` to avoid invoking the function if not necessary, the log message will get dropped down the stack by another check.

It would also be best to keep this pull request focused on just adding the trace log level, so just keep the current code structure with the return statement even though it is a little verbose.

Reply:
Reverted every change not directly related to implementation of Trace log level.